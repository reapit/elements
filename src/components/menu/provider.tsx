import React, { createContext, Dispatch, FC, HTMLAttributes, SetStateAction, useState } from 'react'

interface MenuContextValue {
  getTriggerProps: () => HTMLAttributes<HTMLButtonElement>
  getPopoverProps: () => HTMLAttributes<HTMLDivElement>
  isOpen: boolean
  setIsOpen: Dispatch<SetStateAction<boolean>>
}

const MenuContext = createContext<MenuContextValue | null>(null)

const MenuProvider: FC = ({ children }) => {
  const [isOpen, setIsOpen] = useState(false)

  const getTriggerProps = () => {
    return {
      onClick: () => {
        setIsOpen((prev) => !prev)
      },
      'aria-haspopup': 'true',
      'aria-expanded': isOpen,
      className: 'menu-trigger',
    } as ReturnType<MenuContextValue['getTriggerProps']>
  }

  const getPopoverProps = () => {
    return {
      'aria-labelledby': 'menu-header',
      role: 'menu',
      hidden: !isOpen,
      className: 'menu-popover',
    }
  }

  return (
    <MenuContext.Provider value={{ getTriggerProps, getPopoverProps, isOpen, setIsOpen }}>
      {children}
    </MenuContext.Provider>
  )
}

export { MenuContext, MenuProvider }

export const useMenu = () => {
  const context = React.useContext(MenuContext)
  if (!context) {
    throw new Error('useMenu must be used within a MenuProvider')
  }
  return context
}
