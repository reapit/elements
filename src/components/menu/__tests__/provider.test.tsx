import { fireEvent, render } from '@testing-library/react'
import { MenuProvider, useMenu } from '../provider'
import '@testing-library/jest-dom'

const TestComponent = () => {
  const { getTriggerProps, getPopoverProps, isOpen } = useMenu()

  return (
    <div>
      <button {...getTriggerProps()}>Trigger</button>
      <div {...getPopoverProps()}>{isOpen && <span>Popover Content</span>}</div>
    </div>
  )
}

describe('MenuProvider and useMenu', () => {
  it('should render MenuProvider and match snapshot', () => {
    const { asFragment } = render(
      <MenuProvider>
        <TestComponent />
      </MenuProvider>,
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('should toggle isOpen state when trigger button is clicked', () => {
    const { getByText, queryByText } = render(
      <MenuProvider>
        <TestComponent />
      </MenuProvider>,
    )

    expect(queryByText('Popover Content')).toBeNull()

    fireEvent.click(getByText('Trigger'))

    expect(getByText('Popover Content')).toBeInTheDocument()

    fireEvent.click(getByText('Trigger'))

    expect(queryByText('Popover Content')).toBeNull()
  })

  it('should correctly set aria attributes on trigger and popover elements', () => {
    const { getByText, container } = render(
      <MenuProvider>
        <TestComponent />
      </MenuProvider>,
    )

    const triggerButton = getByText('Trigger')
    expect(triggerButton).toHaveAttribute('aria-haspopup', 'true')
    expect(triggerButton).toHaveAttribute('aria-expanded', 'false')

    fireEvent.click(triggerButton)

    expect(triggerButton).toHaveAttribute('aria-expanded', 'true')

    const popover = container.querySelector('.menu-popover')
    expect(popover).toHaveAttribute('aria-labelledby', 'menu-header')
    expect(popover).toHaveAttribute('role', 'menu')
    expect(popover).not.toHaveAttribute('hidden')

    fireEvent.click(triggerButton)
    expect(popover).toHaveAttribute('hidden')
  })

  it('should throw an error when useMenu is used outside of MenuProvider', () => {
    const TestErrorComponent = () => {
      useMenu()
      return null
    }
    const renderWithoutProvider = () => render(<TestErrorComponent />)

    expect(renderWithoutProvider).toThrow('useMenu must be used within a MenuProvider')
  })
})
