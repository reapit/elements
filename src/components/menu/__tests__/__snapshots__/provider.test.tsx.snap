// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuProvider and useMenu should render MenuProvider and match snapshot 1`] = `
<DocumentFragment>
  <div>
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="menu-trigger"
    >
      Trigger
    </button>
    <div
      aria-labelledby="menu-header"
      class="menu-popover"
      hidden=""
      role="menu"
    />
  </div>
</DocumentFragment>
`;
